base/interactiveutil.jl

```diff
@@ -465,204 +465,84 @@ end
 whos(m::Module, pat::Regex=r"") = whos(STDOUT, m, pat)
 whos(pat::Regex) = whos(STDOUT, current_module(), pat)
 
-"""
-    summarysize(obj, recurse) => Int
-
-summarysize is an estimate of the size of the object
-as if all iterables were allocated inline
-in general, this forms a conservative lower bound
-n the memory "controlled" by the object
-if recurse is true, then simply reachable memory
-should also be included, otherwise, only
-directly used memory should be included
-you should never ignore recurse in cases where recursion is possible"""
-summarysize(obj::ANY, recurse::Bool) = try convert(Int, sizeof(obj)); catch; Core.sizeof(obj); end
-
-# these three cases override the exception that would be thrown by Core.sizeof
-summarysize(obj::Symbol, recurse::Bool) = 0
-summarysize(obj::DataType, recurse::Bool) = 0
-function summarysize(obj::Module, recurse::Bool)
-    size::Int = sizeof(obj)
-    if recurse
-        for binding in names(obj, true)
-            if isdefined(obj, binding)
-                value = getfield(obj, binding)
-                if (value !== obj) # skip the self-recursive definition
-                    recurseok = !isa(value, Module) || module_parent(value) === obj
-                    size += summarysize(value, recurseok)::Int # recurse on anything that isn't a module
-                end
-            end
-        end
-    end
-    return size
-end
+#################################################################################
 
-function summarysize(obj::Task, recurse::Bool)
-    size::Int = sizeof(obj)
-    if recurse
-        if isdefined(obj, :code)
-            size += summarysize(obj.code, true)::Int
-        end
-        size += summarysize(obj.storage, true)::Int
-
-        size += summarysize(obj.backtrace, false)::Int
-        size += summarysize(obj.donenotify, false)::Int
-        size += summarysize(obj.exception, false)::Int
-        size += summarysize(obj.result, false)::Int
-    end
-    return size
-end
+summarysize(obj; exclude = Union{Module,Function}) = summarysize(obj, ObjectIdDict(), exclude)
 
-function summarysize(obj::SimpleVector, recurse::Bool)
-    size::Int = sizeof(obj)
-    if recurse
-        for val in obj
-            if val !== obj
-                size += summarysize(val, false)::Int
-            end
-        end
-    end
-    return size
-end
+# these cases override the exception that would be thrown by Core.sizeof
+summarysize(obj::Symbol, seen, excl) = 0
+summarysize(obj::DataType, seen, excl) = 0
 
-function summarysize(obj::Tuple, recurse::Bool)
-    size::Int = sizeof(obj)
-    if recurse
-        for val in obj
-            if val !== obj && !isbits(val)
-                size += summarysize(val, false)::Int
+function summarysize(obj::ANY, seen, excl)
+    key = pointer_from_objref(obj)
```

Comment:
obj isn't necessarily rooted for the lifetime of seen, so pointer_from_objref isn't unique